FieldOpt/Optimization/optimizers/CMA_ES.h

```diff
@@ -0,0 +1,126 @@
+
+   This file is part of the FieldOpt project.
+
+   FieldOpt is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   FieldOpt is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with FieldOpt.  If not, see <http://www.gnu.org/licenses/>.
+******************************************************************************/
+
+#ifndef FIELDOPT_CMA_ES_H
+#define FIELDOPT_CMA_ES_H
+
+#include <boost/random.hpp>
+#include "optimizer.h"
+
+namespace Optimization {
+namespace Optimizers {
+
+
+/*!
+* @brief This class implements the "Covariance Matrix Adaption Evolutionary Strategy" (CMA-ES) algorithm.
+*
+* It samples the space using a multivariate normal distribution in each real-valued dimension. For each step the
+* variables in the population is perturbed to achieve a zero mean, based on the covariance matrix. For each step
+* the coverance matrix is based on the new distribution around the new mean.
+*
+* It approximates the
+* The implementation is based on the description found at:
+* http://delivery.acm.org/10.1145/3080000/3075986/p183-sakamoto.pdf?ip=129.241.230.160&id=3075986&acc=ACTIVE%20SERVICE&key=CDADA77FFDD8BE08%2E5386D6A7D247483C%2E4D4702B0C3E38B35%2E4D4702B0C3E38B35&__acm__=1560861322_554a2416a12a88617c76b0a945b1687c
+*
+*
+*
+*
+*/
+class CMA_ES : public Optimizer {
+public:
+    CMA_ES(Settings::Optimizer *settings,
+           Case *base_case,
+           Model::Properties::VariablePropertyContainer *variables,
+           Reservoir::Grid::Grid *grid,
+           Logger *logger,
+           CaseHandler *case_handler=0,
+           Constraints::ConstraintHandler *constraint_handler=0);
+protected:
+    void handleEvaluatedCase(Case *c) override;
+    void iterate() override;
+    virtual TerminationCondition IsFinished() override;
+protected:
+    boost::random::mt19937 gen_; //!< Random number generator with the random functions in math.hpp
+    std::mt19937 gen;
+public:
+    struct Individual{
+        Eigen::VectorXd rea_vars_; //!< Real variables
+        Case *case_pointer_; //!< Pointer to the case
+        Eigen::VectorXd erands_norm_; //!< Normalized real variables
+        double penalty_dist_; //!< Sum of normalized distance outside of the "distance"
+        int index_; //!< index
+        Individual(Optimization::Case *c, boost::random::mt19937 &gen, int index, Eigen::VectorXd erands_norm, double penalty_dist);
+        Individual(){}
+        double ofv() { return case_pointer_->objective_function_value(); }
+    };
+
+    Settings::Optimizer *settings_;
```

Comment:
I'm assuming you don't intend these members to be public.
You should add a `private:` above this line.